File: java/src/main/java/com/danifoldi/dml/type/DmlComment.java
package com.danifoldi.dml.type;

import com.danifoldi.dml.DmlSerializer;

public class DmlComment extends DmlValue {
    private String value;

    public DmlComment(String value) {
        this.value = value;
    }

    public String value() {
        return this.value;
    }

    public void value(String value) {
        this.value = value;
    }

    @Override
    public String serialize(int indent) {
        int spaceCount = value.lines().anyMatch(s -> s.startsWith("* ")) ? 1 : 2;
        return DmlSerializer.indent(!value.contains("\n") ? "//" + (value.startsWith(" ") ? "" : " ") + value : "/* " + DmlSerializer.indent(value, spaceCount) + "\n */", indent);
    }

    @Override
    public String toString() {
        return "DmlComment{" + value + "}";
    }
}

File: javascript/src/types/DmlComment.js
import DmlSerializer from '../DmlSerializer.js'

class DmlComment {
  #commentValue

  constructor(commentValue) {
    this.#commentValue = commentValue
  }

  value(commentValue) {
    if (typeof commentValue === 'undefined') {
      return this.#commentValue
    }

    this.#commentValue = commentValue
  }

  serialize(indent) {
    let spaceCount = this.#commentValue.split('\n').some(l => l.trimLeft().startsWith('* ')) ? 1 : 2
    return DmlSerializer.indent(this.#commentValue.indexOf('\n') === -1
      ? '//' + (this.#commentValue.startsWith(' ') ? '' : ' ') + this.#commentValue
      : '/*' + DmlSerializer.indent(this.#commentValue, spaceCount) + '*/', indent)
  }

  get [Symbol.toStringTag]() {
    return `DmlComment{${this.#commentValue}}`
  }
}

export default DmlComment
